ehrql/backends/tpp.py

```diff
@@ -109,9 +109,24 @@ def modify_query_variables(self, variables):
         # return it unmodified
         if self.include_t1oo:
             return variables
-        # Otherwise we add an extra condition to the population definition which is that
-        # the patient does *not* appear in the T1OO table.
-        extra_condition = ~t1oo.exists_for_patient()
+        # Otherwise we add an extra condition to the population definition to include
+        # only patients who we know have not opted out
+        patients = ehrql.tables.tpp.patients
+        registrations = ehrql.tables.tpp.practice_registrations
+        extra_condition = (
+            # Include only patients who do *not* have a T1OO recorded
+            ~t1oo.exists_for_patient()
+            & (
+                # Of those, include only those who have a currently active practice
+                # registration (otherwise they may have recorded a T1OO elsewhere) ...
+                registrations.where(
+                    registrations.end_date.is_null()
+                ).exists_for_patient()
+                # ... or those who have died and therefore cannot have registered a T1OO
+                # elsewhere
+                | patients.date_of_death.is_not_null()
```

Comment:
Just checking - if a patient has a date of death recorded in TPP, then they must have died while registered at a TPP practice. Do we need to worry about possible data entry errors where deregistration date is before the recorded date of death?

Reply:
Data entry errors in the sense of a patient being recorded as having died when in fact they haven't? I don't want to say that's impossible, but there's a certain point at which we just have to rely on the data we're given.

If the worry is about the exact date of death being incorrect then I don't think that matters because, as you say, the fact that death is recorded in TPP indicates they were registered with a SystmOne practice at time of death – and that's all that's relevant for our purposes.